src/program/config/listen/listen.lua

```diff
@@ -0,0 +1,318 @@
+-- Use of this source code is governed by the Apache 2.0 license; see COPYING.
+module(..., package.seeall)
+
+local S = require("syscall")
+local ffi = require("ffi")
+local rpc = require("lib.yang.rpc")
+local yang = require("lib.yang.yang")
+local data = require("lib.yang.data")
+local path_lib = require("lib.yang.path")
+local common = require("program.config.common")
+
+local function buffered_input_from_fd(fd)
+   local buf_size = 4096
+   local buf = ffi.new('uint8_t[?]', buf_size)
+   local buf_end = 0
+   local pos = 0
+   local ret = {}
+   local eof = false
+   local function fill()
+      assert(pos == buf_end)
+      if eof then return 0 end
+      pos = 0
+      buf_end = assert(S.read(fd, buf, buf_size))
+      assert(0 <= buf_end and buf_end <= buf_size)
+      if buf_end == 0 then eof = true end
+      return buf_end
+   end
+   function ret:avail() return buf_end - pos end
+   function ret:getfd() return fd end
+   function ret:eof() return eof end
+   function ret:peek()
+      if pos == buf_end and fill() == 0 then return nil end
+      return string.char(buf[pos])
+   end
+   function ret:discard()
+      assert(pos < buf_end)
+      pos = pos + 1
+   end
+   return ret
+end
+
+local whitespace_pat = '[ \n\r\t]'
+
+local function drop_buffered_whitespace(input)
+   while input:avail() > 0 and input:peek():match(whitespace_pat) do
+      input:discard()
+   end
+end
+
+local function take_while(input, pat)
+   local out = {}
+   while input:peek() and input:peek():match(pat) do
+      table.insert(out, input.peek())
+      input:discard()
+   end
+   return table.concat(out)
+end
+
+local function check(input, ch)
+   if input:peek() ~= ch then return false end
+   input:discard()
+   return true
+end
+
+local function consume(input, ch)
+   if not check(input, ch) then
+      if input:eof() then error('unexpected EOF') end
+      error('expected '..ch..', got '..input:peek())
+   end
+end
+
+local function consume_pat(input, pat)
+   local ch = input:peek()
+   if ch == nil then error('unexpected EOF') end
+   if not ch:match(pat) then error('unexpected character '..ch) end
+   input:discard()
+   return ch
+end
+
+local function skip_whitespace(input) take_while(input, whitespace_pat) end
+
+-- Pattern describing characters that can appear literally in a JSON
+-- string.
+local literal_string_chars_pat = '%w'
+do
+   -- Printable non-alphanumeric ASCII chars, excluding control
+   -- characters, backslash, and double-quote.
+   local punctuation = "!#$%&'()*+,-./:;<=>?@[]^_`{|}~ "
+   for i=1,#punctuation do
+      local punctuation_pat = '%'..punctuation:sub(i,i)
+      literal_string_chars_pat = literal_string_chars_pat..punctuation_pat
+   end
+   literal_string_chars_pat = '['..literal_string_chars_pat..']'
+end
+-- The escapable characters in JSON.
+local escaped_string_chars =
+   { r="\r", n="\n", t="\t", ["\\"]="\\", ['"']='"', b="\b", f="\f", ["/"]="/" }
+
+local function read_json_string(input)
```

Comment:
Might want to leave a comment to say what's supported and isn't for future reference but otherwise the reasoning sounds good.